[jordankanghm] iP

jordankanghm opened the pull request:
### Dookie
DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST TO USE

All you need to do is,

1. download it from [here](https://github.com/jordankanghm/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jordyyyyyyy committed:
Level 0 Commit

jordyyyyyyy committed:
Level 1 Commit

jordyyyyyyy committed:
Level 2 Commit

jordyyyyyyy committed:
Level 2 New Commit

jordyyyyyyy committed:
Level 3 Commit

jordyyyyyyy committed:
Level 3 New Commit

jordyyyyyyy committed:
Level 3 New Commit

jordyyyyyyy committed:
Level 4 Commit

jordyyyyyyy committed:
A-TextUiTesting Commit

jordyyyyyyy committed:
Level 5 Commit

jordyyyyyyy committed:
Level 6 Commit

jordyyyyyyy committed:
Level 4, A-TextUiTesting, 5, 6, A-Enums Commit

jordyyyyyyy committed:
Save tasks when list changes and load data on start.

Currently, tasks added or updated are not saved when the program is terminated.

Tasks must be saved as users may require the list of tasks the next time they use the program.

Let's save accumulated tasks in a duke.txt file automatically whenever the task list changes and load the task list from the file when the chatbot starts up.

Having a specific file to store the task list allows changes to be saved and loaded easily.

jordyyyyyyy committed:
Initial Commit

jordyyyyyyy committed:
Understand Deadlines.

Deadlines are treated as simply strings.

We need to have some form of checking the string input to ensure that a proper date and time is inputted.

Let's verify that the date and time inputted by the user is valid and store it in the Deadline object as a LocalDateTime object.

By doing so, we can be sure that the user has inputted a valid deadline and code is more organised.

jordyyyyyyy committed:
Understand Deadlines.

Deadlines are treated as simply strings.

We need to have some form of checking the string input to ensure that a proper date and time is inputted.

Let's verify that the date and time inputted by the user is valid and store it in the Deadline object as a LocalDateTime object.

By doing so, we can be sure that the user has inputted a valid deadline and code is more organised.

jordyyyyyyy committed:
Merge branch 'branch-Level-7'

jordyyyyyyy committed:
Manually committing to merge with Level-8 branch

jordyyyyyyy committed:
Structure the code using more OOP.

Most of the code is cluttered in the Duke.java file.

This makes the card hard to read, and makes it harder to change when the code is not segmented.

Let's edit the code and use more OOP!

Incorporating more OOP makes the code easier to read and more scalable as related fields and methods are grouped together.

jordyyyyyyy committed:
Merge branch 'branch-A-MoreOOP'

jordyyyyyyy committed:
Encapsulate files in a duke package.

All files are currently in the src/main/java directory.

The files must be encapsulated as it is easier to determine which files are related to which.

Let's place all files in a package and include them in that package.

By putting all files under a package, it encapsulates the Duke chatbot and its functionalities.

jordyyyyyyy committed:
Merge branch 'branch-A-Packages'

jordyyyyyyy committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

jordyyyyyyy committed:
Implement Gradle for Testing.

There is no way of doing automated testing.

Testing our program manually is time consuming and error-prone.

Let's use Gradle to help us automate testing.

Gradle has functions that allow it to test cases automatically.

jordyyyyyyy committed:
Add one more file that was forgotten in the previous commit.

jordyyyyyyy committed:
Merge branch 'branch-A-Gradle'

jordyyyyyyy committed:
Include JUnit Testing.

There are no automated test cases to test the code.

We need automated test cases to determine if our code works and to make the testing process more efficient.

Let's implement some JUnit test cases!

By implementing JUnit test cases, we can test our code more easily and efficiently.

jordyyyyyyy committed:
Merge branch 'branch-A-JUnit'

jordyyyyyyy committed:
Make adjustments for JAR.

The program prints messages which should not be printed, which comes from program testing.

The testing messages must be removed to ensure a seamless experience for users.

Let's remove all testing messages!

By doing so, unnecessary code will not be shown to the users, thus improving their experience.

jordyyyyyyy committed:
Merge branch 'branch-A-Jar'

jordyyyyyyy committed:
Add JavaDocs to the code.

Classes and methods in the code have no descriptions or explanations.

Users or other programmers may not be able to understand the classes and methods used.

Let's use JavaDocs to document our classes and methods!

JavaDocs provides a structured way of documenting code for others to understand.

jordyyyyyyy committed:
Merge branch 'branch-A-JavaDoc'

jordyyyyyyy committed:
Apply Coding Conventions

Currently, the code does not follow a standard convention.

Code can be hard to read if coding styles are not standardised.

Let's use the SE-EDU Java Coding Standard to standardise the code!

The SE-EDU Java Coding Standard provides a clear and concise coding standard which ensures the code is structured and readable.

jordyyyyyyy committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/seedu/duke/TaskList.java

jordyyyyyyy committed:
Correct Mistakes in A-CodingStandard merge

jordyyyyyyy committed:
Add Find Functionality

There is no way to query the task list using a string.

Users cannot find tasks related to their search keywords.

Let's implement a search feature for users to find tasks related to their query!

Users can now search for tasks related to their query.

jordyyyyyyy committed:
Merge branch 'branch-Level-9'

jordyyyyyyy committed:
Add minor improvements in the code

jordyyyyyyy committed:
Testing

jordyyyyyyy committed:
Implement LocalDateTime for Events

The start and end dates inputted into events are not checked whether or not they are actual dates.

This can result in the start and end dates being words instead of dates.

Let's use LocalDateTime for Events as well.

In this way, the start and end dates can be determined efficiently and and effectively.

jordyyyyyyy committed:
Streamline Deadline and Event Dates

Deadline and Event Tasks require / and - for their time inputs respectively.

This can be confusing for the user whether to use / or -.

Let's streamline both Tasks to use -!

/ is not used as it is a requirement for one of the Parser methods.

jordyyyyyyy committed:
Merge branch 'branch-streamline-date-formats'

tjch-o commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    // Checks if an input task has a task title
+    public boolean descriptionIsEmpty(String cmd) {
```
perhaps the order of the words can be rearranged to sound more like a question such as `isDescriptionEmpty`?

tjch-o commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,82 @@
+        ArrayList<Task> tasks = taskList.getTasks();
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = tasks.get(i);
+            System.out.println(i+1 + "." + task.getTaskType() + task.getStatusIcon() + " " + task.name +
```
should there be more spacing for the `i+1`?

jordyyyyyyy committed:
Add JavaFX GUI

The application is currently run on the CLI.

This is unaesthetic and downplays the user experience.

Let's use JavaFX to implement a GUI for the user!

Users can now have a more pleasant experience using the application.

jordyyyyyyy committed:
Merge branch 'branch-Level-10'

tjch-o commented on src/main/java/seedu/duke/Ui.java:
```diff
@@ -0,0 +1,144 @@
+
+        for (int i = 0; i < results.size(); i++) {
+            Task task = results.get(i);
+            message += "     " + (i+1) + "." + task.getTaskType() + task.getStatusIcon() + " " + task.name +
```
same comment as before but perhaps the `i+1` could do with some spacing?

tjch-o commented on src/test/java/seedu/duke/StorageTest.java:
```diff
@@ -0,0 +1,94 @@
+     * Tests the save method in Storage.
+     */
+    @Test
+    public void save_normalTaskArrayList_success(){
```
good naming of the Junit test cases!

tjch-o commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    // Removes all data from the duke.txt file
+    // Takes the task arraylist argument, serializes it and stores it in the duke.txt file
```
I like the explanation but would this be better if it was in a JavaDocs comment?

tjch-o reviewed (commented):
LGTM! generally there is not much coding standard issues, just minor nits to fix.

nicleongyj commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,29 @@
+        this.isDone = true;
+    }
+
+    public void unMarkTask() {
```
maybe this should be unmarkTask() since unmark is a word by itself?

nicleongyj commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    // Prints the message at the end of the program
+    public void printExit() {
```
naming could be printExitMessage() to match the rest of the method names

nicleongyj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    // Takes a command and executes it accordingly
+    public void parse(String cmd) {
```
maybe you can consider encapsulating the error handling in the markOrDelete() method to reduce repeated code and improve readability in this method

nicleongyj reviewed (commented):
Overall good job in maintaining consistent coding standards throughout the project! Keep up the good work :)

jordyyyyyyy committed:
Remove Duplicate Files in Github

There are duplicate files from previous pushes to Github.

It is confusing for users when there are duplicate files as they do not know which files to examine.

Let's remove the duplicate files!

Users will have a clearer image of how the program is run.

jordyyyyyyy committed:
Merge branch 'branch-remove-duplicate-files'

jordyyyyyyy committed:
Standardise Duke replies

Some Duke replies still have dashes which was used in the CLI.

The dashes are unaesthetic and are not consistent.

Let's remove dashes from all Duke replies.

Replies are consistent and aesthetic after the changes.

jordyyyyyyy committed:
Merge branch 'branch-standardise-replies'

jordyyyyyyy committed:
Use Assert Feature

The current code does not have a way to document important assumptions that should hold at various points in the code.

This can make the debugging process more difficult as we do not have hints to where the error might be.

Let's add Java Assert statements to help in our debugging process!

Doing so will make our debugging process more efficient.

jordyyyyyyy committed:
Improve Code Quality

The code is excessively long and does not adhere to the SLAP principle.

This makes the code difficult to read and is unintuitive.

Let's improve the quality of code!

This makes the code more readable and understandable.

jordankanghm committed:
Merge pull request #1 from jordankanghm/branch-A-Assertions

Use Assert Feature

jordyyyyyyy committed:
Merge branch 'master' into branch-A-Code-Quality

# Conflicts:
#	src/main/java/seedu/duke/Parser.java

jordankanghm committed:
Merge pull request #2 from jordankanghm/branch-A-Code-Quality

Branch a code quality

jordyyyyyyy committed:
Add a postpone feature

Currently there is no way to change the deadline of a Deadline task or the duration of a Event task.

Users must delete the current task and add a new task to change its details.

Let's add a feature to allow users to postpone deadlines/durations!

Users can change the details of an event more easily.

jordyyyyyyy committed:
Merge branch 'branch-snooze'

jordyyyyyyy committed:
Merge branch 'master' of https://github.com/jordankanghm/ip

jordyyyyyyy committed:
Improve Code Quality

Currently, the naming of methods in the Ui class are misleading and the conditional statements in the Parser class are hard to read.

This can cause others to have a hard time reading the code.

Let's improve the naming conventions and use switch case statements instead!

This makes it easier for others to read the code.

jordyyyyyyy committed:
Resolve issues regarding invalid task type

jordyyyyyyy committed:
Rename the chatbot to Dookie

jordyyyyyyy committed:
Improve Code Quality

jordyyyyyyy committed:
Improve Test Cases

jordyyyyyyy committed:
Add the product screenshot to the docs folder

jordyyyyyyy committed:
Add title to GUI window

jordyyyyyyy committed:
Create the User Guide

jordyyyyyyy committed:
Create new classes for commands

Currently, the Parser class is extremely long as most of the command methods
are contained within the class.

This makes the code extremely long and probes further abstraction and
simplification of the code.

Let's make new classes to encapsulate the commands!

This way, our code has a higher level of abstraction and is more organised.
